### ISSUE TITLE
Evaluating the uncertainty of sdf-based NeRFs

### ISSUE THREAD

**Issue by Anonymous772066235** [BayesRays/BayesRays#7]
Hi!

Thank you for the work. Could you please tell me if this work can be migrated to sdfstudio to evaluate the mesh of sdf-based NeRFs(such as neus,volsdf...)? I skimmed the paper and the code, and here is my understanding: construct an implicit uncertainty field using multiple viewpoints, and then input a 3D point to determine the uncertainty at the point. If I understand correctly, discretizing the mesh or sdf as the points cloud would similarly query their uncertainty. Is this feasible?

Thank you.

**Comment by hariharan1412** [comment]
Hi, did installation go successfully for you? I'm facing some issues with running ns-uncertinity.

**Comment by lilygoli** [comment]
@Anonymous772066235 
Hi, 

the work has not been tested on sdf-based representations but it should be able to generalize to for example neus. However, you should adopt the code to work in sdf studio, as the uncertainty extraction code in this repository is not completely independent of nerfstudio code. As an example you can look at the uncertainty.py file to see how the same uncertainty extraction has been added to the three different nerf codes (ingp, nerfacto and mipnerf).
Once you have it on your sdf-based representation you will have a 3D field you can query anywhere in space for the uncertainty measure.

**Comment by Anonymous772066235** [comment]
> @Anonymous772066235 Hi,
> 
> the work has not been tested on sdf-based representations but it should be able to generalize to representations like neus. However, you should adopt the code to work in sdf studio, as the uncertainty extraction code in this repository is not completely independent of nerfstudio code. As an example you can look at the uncertainty.py file to see how the same uncertainty extraction has been added to the three different nerf codes (ingp, nerfacto and mipnerf). Once you have it on your sdf-based representation you will have a 3D field you can query anywhere in space for the uncertainty measure.

Ok, thank you！I will try.

**Comment by Anonymous772066235** [comment]
> Hi, did installation go successfully for you? I'm facing some issues with running ns-uncertinity.

Not successful. I am going to create a docker container based on nerfstudio’s docker image.

**Comment by lilygoli** [comment]
what issue did you run into for installation?

**Comment by Anonymous772066235** [comment]
@lilygoli 
Hi,

I have recently successfully installed BayesRays and completed the evaluation of nerfacto results. That's awsome! Referring to your implementation of `get_unc_mipnerf`, I successfully implemented uncertainty evaluation and visualization for `vanilla-nerf`. However, I encountered an issue when migrating to the `sdf-based` method. I tried both `bakedsdf` and `neus-facto` under `sdfstudio`, as well as `neus` under `nerfstudio`, but they all presented the same problem: in the `find_uncertainty` method, `defform_points.grad` is `None`. Could you provide some guidance on this?

![image](https://github.com/BayesRays/BayesRays/assets/58173016/5b37ef69-ab23-4953-bf96-9e11d6d13b34)
Screenshots of successful implementation of  `vanilla-nerf `.

![image](https://github.com/BayesRays/BayesRays/assets/58173016/6eaed859-a1e6-49a4-b1f4-c93baee390e3)
Screenshot of the error encountered while running the  `sdf-based` experiment.

This is the code uesed for uncertainty evaluation of  `neus`:
```python
from nerfstudio.models.neus import NeuSModel
from typing import  Dict, List, cast
from torch.nn import Parameter
......
    def get_unc_neus(self, ray_bundle,model) :
        if model.collider is not None:
            ray_bundle = model.collider(ray_bundle)

        assert (
            ray_bundle.metadata is not None and "directions_norm" in ray_bundle.metadata
        ), "directions_norm is required in ray_bundle.metadata"

        samples_and_field_outputs = model.sample_and_forward_field(ray_bundle=ray_bundle)

        # shortcuts
        field_outputs: Dict[FieldHeadNames, torch.Tensor] = cast(
            Dict[FieldHeadNames, torch.Tensor], samples_and_field_outputs["field_outputs"]
        )
        ray_samples = samples_and_field_outputs["ray_samples"]
        weights = samples_and_field_outputs["weights"]
        bg_transmittance = samples_and_field_outputs["bg_transmittance"]

        points = ray_samples.frustums.get_positions()                                               # add
        pos, _ = normalize_point_coords(points, self.aabb, model.field.spatial_distortion)          # add
        offsets = self.deform_field(pos).clone().detach()                                           # add
        offsets.requires_grad = True                                                                # add
        ray_samples.frustums.set_offsets(offsets)                                                   # add

        rgb = model.renderer_rgb(rgb=field_outputs[FieldHeadNames.RGB], weights=weights)
        depth = model.renderer_depth(weights=weights, ray_samples=ray_samples)
        # the rendered depth is point-to-point distance and we should convert to depth
        depth = depth / ray_bundle.metadata["directions_norm"]

        normal = model.renderer_normal(semantics=field_outputs[FieldHeadNames.NORMALS], weights=weights)
        accumulation = model.renderer_accumulation(weights=weights)

        # background model
        if model.config.background_model != "none":
            assert isinstance(model.field_background, torch.nn.Module), "field_background should be a module"
            assert ray_bundle.fars is not None, "fars is required in ray_bundle"
            # sample inversely from far to 1000 and points and forward the bg model
            ray_bundle.nears = ray_bundle.fars
            assert ray_bundle.fars is not None
            ray_bundle.fars = torch.ones_like(ray_bundle.fars) * model.config.far_plane_bg

            ray_samples_bg = model.sampler_bg(ray_bundle)
            # use the same background model for both density field and occupancy field
            assert not isinstance(model.field_background, Parameter)
            field_outputs_bg = model.field_background(ray_samples_bg)
            weights_bg = ray_samples_bg.get_weights(field_outputs_bg[FieldHeadNames.DENSITY])

            rgb_bg = model.renderer_rgb(rgb=field_outputs_bg[FieldHeadNames.RGB], weights=weights_bg)
            depth_bg = model.renderer_depth(weights=weights_bg, ray_samples=ray_samples_bg)
            accumulation_bg = model.renderer_accumulation(weights=weights_bg)

            # merge background color to forgound color
            rgb = rgb + bg_transmittance * rgb_bg

            bg_outputs = {
                "bg_rgb": rgb_bg,
                "bg_accumulation": accumulation_bg,
                "bg_depth": depth_bg,
                "bg_weights": weights_bg,
            }
        else:
            bg_outputs = {}

        outputs = {
            "rgb": rgb,
            "accumulation": accumulation,
            "depth": depth,
            "normal": normal,
            "weights": weights,
            # used to scale z_vals for free space and sdf loss
            "directions_norm": ray_bundle.metadata["directions_norm"],
        }
        outputs.update(bg_outputs)

        if model.training:
            grad_points = field_outputs[FieldHeadNames.GRADIENT]
            outputs.update({"eik_grad": grad_points})
            outputs.update(samples_and_field_outputs)

        if "weights_list" in samples_and_field_outputs:
            weights_list = cast(List[torch.Tensor], samples_and_field_outputs["weights_list"])
            ray_samples_list = cast(List[torch.Tensor], samples_and_field_outputs["ray_samples_list"])

            for i in range(len(weights_list) - 1):
                outputs[f"prop_depth_{i}"] = model.renderer_depth(
                    weights=weights_list[i], ray_samples=ray_samples_list[i]
                )
        # this is used only in viewer
        outputs["normal_vis"] = (outputs["normal"] + 1.0) / 2.0
        return outputs, points, offsets
```



The  code contains added statements related to `points` and `offsets`  based  on `get_output` method in `base_surface_model.py`. From my observation, the `get_unc_model` method is built upon the `get_output` method of relevant models. Moreover, I've successfully performed evaluations for `vanilla-nerf` using this approach. However, the transition to `sdf-based` methods hasn't been successful. I'd greatly appreciate your guidance on this matter.

**Comment by Shrisudhan** [comment]
Hi @Anonymous772066235,
What is the type of normal you are computing in the line:
`normal = model.renderer_normal(semantics=field_outputs[FieldHeadNames.NORMALS], weights=weights)`
Numerical normal by itself doesn't compute gradients for the normals. But analytical normal compute gradients wrt to inputs. So, you might want to enable gradient propagation through your input coordinates and then compute backward to input coordinates...

**Comment by lilygoli** [comment]
@Anonymous772066235 
Hey, 

Happy you were able to test on vanilla nerf. I would make sure that you have the gradients wrt to the positions (even when deformation is not added, and in the original pipeline) available. This depends on how the accelaration backend (tinycudnn or nerfacc) is handled and if its allowing gradient flow all the way back to the point coordinates. Also make sure that the second-order derivatives that might show up because of NeUS formulation are precise enough or you might need to increase precision (for when you fix the gradient flow and stop getting NaNs). Let me know if this works out!

BTW you mentioned you might make a docker container for the repo, if you ended up doing that, would you mind sharing the image?

Thanks!

**Comment by Anonymous772066235** [comment]
@Shrisudhan @lilygoli 
Thanks for your replies. I'll try more.  Actually the docker environment creation failed, and I'm currently testing it in my local conda environment.

**Comment by Anonymous772066235** [comment]
@lilygoli 
Hi,

Due to a lack of theoretical knowledge, some of my attempts haven't been successful yet. Now I can confirm that in   **_find_uncertainty,_**  variable `defom_points` is all zeros and its gradient  is `None` before the line `colors[0].backward(retain_graph=True)` . After that line, `defom_points `remains zero, but its gradient is no longer `None`. However, with the SDF-based NeRF, after that line, `defom_points` remains zero, and its gradient remains 'None'. This is very puzzling for me. I noticed that `nerfstudio` includes methods like `neus` and `neus-facto.` Could you  implement their uncertainty evaluation?

**Comment by lilygoli** [comment]
Hi,

sorry for the late reply. I think your observation is consistent with the fact that the backend is not using gradient propagation to the input point positions for your case. I will try to take a look and see if I can set up an example with NeUS in nerfstudio this weekend and let you know.

**Comment by Anonymous772066235** [comment]
@lilygoli 
Hi,

Is there any good news? Looking forward to it working on NeUS.
